### PR TITLE
chore: improve token type guards part 2

### PR DESCRIPTION
**Pull request by armano2**
This is follow up PR that adds missing type guards for negated functions and adds new union type `TokenOrComment` (moved out from indent rule)

### CODE REVIEW

File: packages/eslint-plugin/src/util/astUtils.ts [hunk 1]
```diff
@@ -4,27 +4,29 @@ import {
   TSESTree,
 } from '@typescript-eslint/experimental-utils';
 
+type NegateGuard<T, U> = T extends U ? T : U;
```
Comment (armano2) [inline packages/eslint-plugin/src/util/astUtils.ts]: i don't like this name, and this is not exclude or extract but it works in similar way

**Comment by bradzacher** [comment]
I don't think this works correctly

image

**Comment by armano2** [comment]
you are correct, this is not going to work, we have to wait for microsoft/TypeScript#29317

**Comment by bradzacher** [comment]
It'd be fine if we explicitly typed `PunctuatorToken.value` with a union of the strings. Then you could just do this and exclude the token from the union.

```ts
type Negated = TSESTree.PuncuatorToken & { value: Exclude<TSESTree.PuncuatorToken['value'], '=>'> };
```

But then we would have to enumerate every single token, which would be a bit of a pain. Though we would only have to do it once.

**Comment by armano2** [comment]
note: there is total of 64 known punctuators in current version of ts

**Comment by bradzacher** [comment]
So you're saying it's a definably finite set? 😛 

If you're happy to go through the painful task of defining the union, I'm happy with it being in here.
Completely up to you if you want to go to that effort!
Depends how much you care about the "correctness" of the typeguards heh

**Comment by armano2** [comment]
This still is not going to work as it should even with types defined :( 
```
    | '{' // OpenBraceToken = 1
    | '}' // CloseBraceToken = 19
    | '(' // OpenParenToken = 20
    | ')' // CloseParenToken = 21
    | '[' // OpenBracketToken = 22
    | ']' // CloseBracketToken = 23
    | '.' // DotToken = 24
    | '...' // DotDotDotToken = 25
    | ';' // SemicolonToken = 26
    | ',' // CommaToken = 27
    | '?.' // QuestionDotToken = 28
    | '<' // LessThanToken = 29
    | '</' // LessThanSlashToken = 30
    | '>' // GreaterThanToken = 31
    | '<=' // LessThanEqualsToken = 32
    | '>=' // GreaterThanEqualsToken = 33
    | '==' // EqualsEqualsToken = 34
    | '!=' // ExclamationEqualsToken = 35
    | '===' // EqualsEqualsEqualsToken = 36
    | '!==' // ExclamationEqualsEqualsToken = 37
    | '=>' // EqualsGreaterThanToken = 38
    | '+' // PlusToken = 39
    | '-' // MinusToken = 40
    | '*' // AsteriskToken = 41
    | '**' // AsteriskAsteriskToken = 42
    | '/' // SlashToken = 43
    | '%' // PercentToken = 44
    | '++' // PlusPlusToken = 45
    | '--' // MinusMinusToken = 46
    | '<<' // LessThanLessThanToken = 47
    | '>>' // GreaterThanGreaterThanToken = 48
    | '>>>' // GreaterThanGreaterThanGreaterThanToken = 49
    | '&' // AmpersandToken = 50
    | '|' // BarToken = 51
    | '^' // CaretToken = 52
    | '!' // ExclamationToken = 53
    | '~' // TildeToken = 54
    | '&&' // AmpersandAmpersandToken = 55
    | '||' // BarBarToken = 56
    | '?' // QuestionToken = 57
    | ':' // ColonToken = 58
    | '@' // AtToken = 59
    | '??' // QuestionQuestionToken = 60
    | '`' // BacktickToken = 61
    | '=' // EqualsToken = 62
    | '+=' // PlusEqualsToken = 63
    | '-=' // MinusEqualsToken = 64
    | '*=' // AsteriskEqualsToken = 65
    | '**=' // AsteriskAsteriskEqualsToken = 66
    | '/=' // SlashEqualsToken = 67
    | '%=' // PercentEqualsToken = 68
    | '<<=' // LessThanLessThanEqualsToken = 69
    | '>>=' // GreaterThanGreaterThanEqualsToken = 70
    | '>>>=' // GreaterThanGreaterThanGreaterThanEqualsToken = 71
    | '&=' // AmpersandEqualsToken = 72
    | '|=' // BarEqualsToken = 73
    | '^='; // CaretEqualsToken = 74
```

**Comment by bradzacher** [comment]
why not? It looks like it should work in the [playground](https://www.typescriptlang.org/play/#code/C4TwDgpgBAKg9gawgOxuaBeK3sB8oDkA3gVAPRlQDykyAQgE4CGAxhPEslFgIwAcAKBxR8BAL6kKUAMIAbOAGcIjVu0QpuUHgE4hOUQApJlGigAKTBig4asAJgAMevIQCUxmfKUWrqdV3seZxFCAG0PU3pmFiRgGwCoOztg0QBdDzlFZWjY+M07AGYUwgA6DwARODj-fIAWYoISpoqqyuA2vPsAVgaAbg8AZQgAWwBLFjh5P058gDYGgBoMuGHhpk7EgHYGgH4y8koARQBXCAVgUbhkDpr7QWFRAB4PABkzhRgACyZp20TdB6ER5kV7vL4-AayJgKT4bApOQEEAB8HgA4lYmMAIAxwb8EgUgojHhhQQoPt9kABRACOxyYsg+tygBWSiKRJIOUHREEx2NxNLpDLhRURGA5UgF9IUkqFTIK9URAEJxZRKQAPFhQtYXK4yxkzLAFHqisUePXm2lSuHzJWmznqzVMbWXKmWhl6uHbUUo+1uhTc3k4ilw+76QgAag8Zlkx31fwKALDBAAtB4ALKjZCxja1BFJgBUHgAgudsaMFAgc4SC4XOSWsQxywh62WKznWUmQZzIdDYUzaiKkwBSKPYtjIaoGqC1BVJ8ORznR2NLuMJWrGpPJ1OcjNZhS77P9m1Jx7PTlvMm4i-kn45r1JpE+qQBhu4l984P90MuZGPtEY18KXfIMfmA3Ec0TH8ADJi2GSAGAUH4ABMNi6PMf1wDw6EsVDqx-AA9DJLAgSc-i6Dsf0VM0NS1TEXVQwcfwAPw8GBRlkJC1CnLpZ2gmC6zg7FEOQJCi0EhDkNQjcMMwzlsIYeTUOPH8dg8E4zh1PFNC6e8fwALmWKZUO-EICAAAWLUiEi6SDTJ2VTOXU84XSczSNlmdDTIAAyw1gEAuGJ3Lw0yVSgD0mVmCjTPDUKV3CqdZkY0zk1Cg9pT9dzeNM-NQpbRsK3iv5Zmk7KcuLUt8ubCqm0KhJZmU0yyFCnsYVqzRZl00yh1iscUGANqsFmEynmJUkb2Qa9+QyiLbNER9QrAoCAI-H4BqgTZPLm+b-x5QDQOWkDkEW1bpqnTZgtEKDcvE4SkLWzYotEXBQvk+6ktEfCSV6TlpGI-rTr+TZal6AQBEzBsADNVCgMxjmQFhgDpYA4BxGoiGCAA3elTj02B-DQSAQbEUHOMdKwoAmZBzigNVcdh+HEcxFH4hB0moXJiG4YRl0oHLAA5KoiwYBg4AAd3iAxkc4OmucZ5HUc4VxcaljRyxh2WkeZmooKgIgoCxmMIFxh0Y04x56YRzWFZQcIDdOAhUgWQgMBRJEoDEEGBDVEo7cwMVnZRb6pAFqBsRFhgwYhqADH5wXhbFiW1VcVxdeCb3fe4f2CBdggg9VeOGFgT41cppDRk0qBRfY2QoHpUWmBABQoCsRGGC4AgoYZCBSAUTM2CgYBPmgUBICbpoSlrkTCAAIhd6fSG+DHoGQOAoDgJeGChMASgEYmgA)

```ts

interface PunctuatorToken {
    value: TokenType;
}

declare const x: PunctuatorToken;
declare function isNotArrowToken(token: PunctuatorToken): token is PunctuatorToken & {
  value: Exclude<PunctuatorToken['value'], '=>'>;
};

x.value === '=>'; // No error
if (isNotArrowToken(x)) {
    x.value === '=>'; // Error: This condition will always return 'false' since the types ... and '"=>"' have no overlap.
}
```

**Comment by armano2** [comment]
its working only if token is `PunctuatorToken` not union of types

```ts
enum TokenType {
    P = 'P',
    B = 'B'
}

type TokenValue =
    | '{' // OpenBraceToken = 18
    | '}' // CloseBraceToken = 19
    | '('; // OpenParenToken = 20

interface PunctuatorToken {
    type: TokenType.B
    value: TokenValue;
}

interface AnyToken {
    type: TokenType.P
    value: string;
}

type UnionTokens = PunctuatorToken | AnyToken

declare const x: PunctuatorToken;
declare function isNotArrowToken(token: UnionTokens):
    token is PunctuatorToken & { value: Exclude<PunctuatorToken['value'], '=>'> };

if (isNotArrowToken(x)) {
    // This condition will always return 'false' since the types 'TokenValue' and '"=>"' 
    x.value === '=>';
} else {
    // value is never
    x.value === '=>';
}

```

[Playground Link](https://www.typescriptlang.org/play/?ssl=34&ssc=1&pln=1&pc=1#code/KYOwrgtgBAKg9ga1DAngB2FA3gKCvqABSgF4oByQ8gGjwICFSL7ycBfHHAF3U3iRAA1AIYAbMJhJ18AHwpZyUAPRKoAeQwh6AJ2EBjYP1BMAjAA5pUOeTaKVUAMKi4AZ2A79hxMbImAnJbWABTkANzKqhqghMLayN4gTABMAAycAJYgXMDaAGaeRGAgelxgwlxw2kaJuARQPBgAXLAJqBgAdPSWAG5iEs3VIuLAoewZWTn5BlAAgiAo1diWDcADrbzthD19q1AuXNqZAOajHNy8UACqIOlwINUuTIRFJWUVVQlWs-PVnAAmwD0olimD0d32UAAHs1nsVSuVKtVRgCgSCoLkXlxbol0i4AHJwLgzbTaOAAd2qQQqAma12xDwAlI1lp9cYU4W9EZ8AGTYKC9YbNACikKBYABAB5Ya8ER8BABtcgCiTkAC61AoJAAfOQtVA2KMcOlclAgriCUSSeTKZCGQylnV7DAABZssEgP7pLF3KBk9KiURQMRk4QoR5xUraRLkfKiNyKFyZaZcZ2YFaPciDHaKYQeigAIm1+cUlkh7WVkhIZHI2rC7CgwDjmFqBHsFagbJAwG6OVL5Z2pCrmp1pxwQA)

**Comment by armano2** [comment]
```ts
enum TokenType {
  Punctuator = 'Punctuator',
  Any = 'Any',
}

type TokenValue = '{' | '}' | '(' | '=>';

interface PunctuatorToken<T extends TokenValue = TokenValue> {
  type: TokenType.Punctuator;
  value: T;
}

interface AnyToken {
  type: TokenType.Any;
  value: string;
}

type UnionTokens =
  | PunctuatorToken<'{'>
  | PunctuatorToken<'}'>
  | PunctuatorToken<'('>
  | PunctuatorToken<'=>'>
  | AnyToken;

declare const x: UnionTokens;
declare function isNotArrowToken<T extends UnionTokens>(
  token: T,
): token is Exclude<T, PunctuatorToken<'{'>>;

if (isNotArrowToken(x)) {
  x.value === '{';                  // string       = '{'
  x.value === '=>';                 // string       = '=>'
  x.type === TokenType.Any;         // TokenType    = TokenType.Any
  x.type === TokenType.Punctuator;  // TokenType    = TokenType.Punctuator
} else {
  x.type === TokenType.Any;         // TokenType.P != TokenType.Any
  x.type === TokenType.Punctuator;  // TokenType.P  = TokenType.P
  x.value === '=>';                 // '{'         != '=>'
  x.value === '{';                  // '{'          = '{'
}
```

Update: defining it in this way is working but Union will have to be build out of `PunctuatorToken` with values
[Playground Link](https://www.typescriptlang.org/play/?ssl=32&ssc=56&pln=32&pc=53#code/KYOwrgtgBAKg9ga1DAngB2FA3gKClABTBAGMAXMAQzLgCcoBeKAciNIuruYBo8oBBECkYtBKHjgC+OHGXSZ4SEADVKAGzCYmzLMygAfFpL2HmAChMsGAPmYBuGQEsQZYLQBmlEpjbkqNWkVQAB4YKGAAD1cQABMAZ1hEUFUNLUSlFM1rbD45DAAudOR5ADpfDgCHfAA3dU1CmAdpHGdXDy9MMSCQHPw84AakkFQMErEqqFrUwriyWmcAcyaZfqgAVRBHOGGhhIY+Q3L-Om7gnWZrA8JiP05AobPjS-xDm4qTh-MLq6O70+YbN8XgIhN0HDgYsASGpKLRMCRtrMoBFChstjslHEHJDobDMO43uioI44gA5OBkfi0WhwADupzCkWi8XWm223Ti1jMuSGDV4AEpCjQlMSEgBRCLQsCQ0Lca7sY73JRnXTWazgxzuKBmEnkynUundMwRfn83rIkpTTSMBjaXR2fCOp3Op0AeldUFm8xACxdInOfAilrqWltVlsDr9Uag7s9c0Wfu0gMDJVWtqY3RGwDGQkjftjmfkTozQyzOZQKbTYcLo1+lXwBdLRcdJaUZbrdCk4TUcUwuHwQarreKo3G0cbbdKBCgAEJh8NSmJK0X00UF7W3orIxOR9npyIa3uU1bQ0mI9Go7HztG5+HmMeQza7fYL5ePdfo8+pDggA)

this is overkill and we should just leave boolean for now

**Pull request closed without merging**